sql/select.go

```diff
@@ -85,3 +99,535 @@ func (p *planner) expandSubqueries(stmt parser.Statement) error {
+// available that can fulfill the query with a more restrictive scan.
+//
+// The current occurs in two passes. The first pass performs a limited form of
+// value range propagation for the qvalues (i.e. the columns). The second pass
+// takes the value range information and determines which indexes can fulfill
+// the query (we currently only support covering indexes) and selects the
+// "best" index from that set. The cost model based on keys per row, key size
+// and number of index elements used.
+func (p *planner) selectIndex(s *scanNode) (planNode, error) {
+	if s.desc == nil || s.filter == nil {
+		// No table or where-clause.
+		return s, nil
+	}
+
+	// Analyze the filter expression to compute the qvalue range info.
+	rangeInfo := make(qvalueRangeMap)
+	rangeInfo.analyzeExpr(s.filter)
+
+	candidates := make([]*indexInfo, 0, len(s.desc.Indexes)+1)
+
+	if s.isSecondaryIndex {
+		// An explicit secondary index was requested. Only add it to the candidate
+		// indexes list.
+		candidates = append(candidates, &indexInfo{
+			desc:  s.desc,
+			index: s.index,
+		})
+	} else {
+		candidates = append(candidates, &indexInfo{
+			desc:  s.desc,
+			index: &s.desc.PrimaryIndex,
+		})
+		for i := range s.desc.Indexes {
+			candidates = append(candidates, &indexInfo{
+				desc:  s.desc,
+				index: &s.desc.Indexes[i],
+			})
+		}
+	}
+
+	for _, c := range candidates {
+		c.analyzeRanges(s, rangeInfo)
+	}
+
+	sort.Sort(indexInfoByCost(candidates))
+
+	if log.V(2) {
+		for i, c := range candidates {
+			log.Infof("%d: selectIndex(%s): %v %s %s",
+				i, c.index.Name, c.cost, c.makeStartKey(), c.makeEndKey())
+		}
+	}
+
+	// After sorting, candidates[0] contains the best index. Copy its info into
+	// the scanNode.
+	s.index = candidates[0].index
+	s.isSecondaryIndex = (s.index != &s.desc.PrimaryIndex)
+	s.startKey = candidates[0].makeStartKey()
+	s.endKey = candidates[0].makeEndKey()
+	return s, nil
+}
+
+// qvalueInfo contains one end of a value range. Op is required to be either
+// GE, GT, LE or LT.
+type qvalueInfo struct {
+	datum parser.Datum
+	op    parser.ComparisonOp
+}
+
+func (q *qvalueInfo) assertStartOp() {
+	switch q.op {
+	case parser.GE, parser.GT:
+	default:
+		panic(fmt.Sprintf("unexpected start op: %s", q.op))
+	}
+}
+
+func (q *qvalueInfo) assertEndOp() {
+	switch q.op {
+	case parser.LE, parser.LT:
+	default:
+		panic(fmt.Sprintf("unexpected end op: %s", q.op))
+	}
+}
+
+func (q *qvalueInfo) compare(datum parser.Datum) int {
+	// TODO(pmattis): We should really be comparing the datums directly, but this
+	// is convenient for now.
+	oldKey, err := encodeTableKey(nil, q.datum)
+	if err != nil {
+		panic(err)
+	}
+	newKey, err := encodeTableKey(nil, datum)
+	if err != nil {
+		panic(err)
+	}
+	return bytes.Compare(newKey, oldKey)
+}
+
+// intersect intersects a new value constraint with an existing constraint. The
+// start parameter controls whether we intersect less than or greater
+// than. Some examples:
```

Comment:
last nit - add that `start == true` corresponds to intersecting `less than` please.